### PR TITLE
Implementation of trainer class (fp16 training & gradient accumulation)

### PR DESCRIPTION
**Pull request by MarcusLoppe**
Hi,

I've created a trainer class that has support for:

- Accelerate which enables mixed precision training (fp16, bf16 or fp8).
- Gradient Accumulation, which enables you to train faster and with a bigger effective batch size.
- Train using total of epochs or steps
- Checkpoints every X epoch/steps
- Evaluation every X steps

I used the trainer in https://github.com/lucidrains/meshgpt-pytorch as a boilerplate, so credits to @lucidrains for the nicely structured trainer.

### CODE REVIEW

File: gliner/trainer.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        epoch_size = len(self.train_loader)
+        self.model.train()  
+        
+        self.scheduler = get_cosine_schedule_with_warmup(
+            self.optimizer,
+            num_warmup_steps=num_warmup_steps,
+            num_training_steps=total_steps
+        )
+        
+        for epoch in range(num_epochs):
```
Comment (urchade) [inline gliner/trainer.py:208]: Adding `.zero_grad()` at the start of the training loop prevent some bug for me.

File: gliner/trainer.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+        if not freeze_token_rep:
+            # If token_rep_layer should not be frozen, add it to the optimizer with its learning rate
+            param_groups.append({"params": model.token_rep_layer.parameters(), "lr": lr_encoder})
+        else:
+            # If token_rep_layer should be frozen, explicitly set requires_grad to False for its parameters
+            for param in self.token_rep_layer.parameters():
+                param.requires_grad = False
+                
+        self.optimizer = torch.optim.AdamW(param_groups, **optimizer_kwargs)
```
[the remark this answers is not in the document]
Comment (MarcusLoppe) [reply gliner/trainer.py:94]: > you can use directly `model.get_optimizer(self, lr_encoder, lr_others, freeze_token_rep=False)` instead of line `80-94`

The reason why I didn't was due to the optimizer_kwargs which is needed for fp16 training. 
I probably can just change the model.get_optimizer function to include a optimizer_kwargs param.

[the remark this answers is not in the document]
Comment (urchade) [reply gliner/trainer.py]: ok, I see

[the remark this answers is not in the document]
Comment (MarcusLoppe) [reply gliner/trainer.py]: > ok, I see

Changed it per suggestion: https://github.com/urchade/GLiNER/pull/85/commits/89803a433e094a59233d1cde5127b5ea9d3bb4aa

**Review by urchade — APPROVED** [review]

**Review by urchade — APPROVED** [review]

**Review by urchade — APPROVED** [review]

**Review by urchade — APPROVED** [review]
